### ISSUE TITLE
How to prevent pod from restarting on PHP errors

### ISSUE THREAD

**Issue by momon** [bitnami/bitnami-docker-wordpress#198]
Howdy folks, while in development it is often I encounter PHP issues that trigger errors.  When the pod starts up or restarts if there is an error, it will cause a reboot on the pod which leads to endless restarts.

I'm able to recover from this more often than not but not without struggling and having to modify the deployment to handle the broken code.  So I've been looking at a command to prevent the pod from restarting when there are PHP errors during boot up but it appears Google does not allow for these directives as explained here:

https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#restart-policy

Here is the Google error when using this directive:

```
Deployment.apps "dpl-test1-fsb-fyxve" is invalid: spec.template.spec.restartPolicy: Unsupported value: "Never": supported values: "Always"
```

Some of the instances when these issues happen are "broken themes" or users who accidentally save function.php file with errors.  If any of these are present when the pod restarts, the pod will read an error and will not bring the website up, resulting in an endless reboot.

Is there any way to prevent these restarts (or have the pod ignore them) during boot up?

Thanks!

**Comment by sameersbn** [comment]
@momon if you are in development mode you should consider disabling the livenessProbe. The linked PR adds chart parameters to disable the liveness and readiness probes

**Comment by momon** [comment]
@sameersbn much appreciate the contribution, however I seem a bit lost on how to make use of it exactly.  Could you clarify what I need to do on my development environment exactly?

**Comment by rafariossaa** [comment]
Hi @momon ,
  In order to disable the livenessProbe, you need to set it when running the helm command
  For example:
`helm install --name my-release stable/wordpress --set readinessProbe.enabled=false`

**Comment by momon** [comment]
Ok, that explains my confusion because I am not using helm.

 I am using a regular deployment.

Can the same command be used as an annotation in the yaml file instead?

**Comment by sameersbn** [comment]
@momon in that case, your deployment yaml would have a `livenessProbe` field. You can comment it out.

**Comment by momon** [comment]
Much appreciate the response @sameersbn but that's not the case here, so I don't know if I am missing something obvious.

Checking one of my live deployments here is the yaml file (I have removed sensitive data):

```
apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "1"
  creationTimestamp: "2019-09-11T04:46:22Z"
  generation: 1
  labels:
    app: ***
    fpc-cust-id: "*"
    fpc-cust-wp-id: "*"
    fpc-cust-wp-label: ****
    fpc-env: customer
    fpc-url-cname: ***
    fpc-url-live: ***
    fpc-url-secret: ****
  name: ****
  namespace: default
  resourceVersion: "161342271"
  selfLink: /apis/apps/v1/namespaces/default/deployments/****
  uid: ****
spec:
  progressDeadlineSeconds: 600
  replicas: 1
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app: ****
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: ****
    spec:
      containers:
      - env:
******
        ports:
        - containerPort: 80
          name: wordpress
          protocol: TCP
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /bitnami
          name: *****
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      terminationGracePeriodSeconds: 30
      volumes:
      - name: *****
        persistentVolumeClaim:
          claimName: *****
status:
  availableReplicas: 1
  conditions:
  - lastTransitionTime: "2019-09-11T04:46:22Z"
    lastUpdateTime: "2019-09-11T04:46:35Z"
    message: ReplicaSet "*****" has successfully progressed.
    reason: NewReplicaSetAvailable
    status: "True"
    type: Progressing
  - lastTransitionTime: "2019-12-19T18:19:59Z"
    lastUpdateTime: "2019-12-19T18:19:59Z"
    message: Deployment has minimum availability.
    reason: MinimumReplicasAvailable
    status: "True"
    type: Available
  observedGeneration: 1
  readyReplicas: 1
  replicas: 1
  updatedReplicas: 1
```

**Comment by sameersbn** [comment]
@momon the yaml file you shared does not show that healthchecks are in use, which means that it is not the reason why the container is getting restarted. I don't see the `image` tag as well, I hope are using an unmodified version of the `bitnami/wordpress` image.

Another reason the container can get restarted is if the main process terminates for some reason, which in this case the `httpd` process. I don't any reason for `httpd` to terminate of php errors.

I request you to please share more information on how you have deployed wordpress and share a scenario under which the issue can be reproduced. It's very hard for us to find a solution to a problem we are not able to understand and/or reproduce.

**Comment by momon** [comment]
My apologies for the lack of details.

I am using the bitnami/wordpress but pulling it as the base for a custom repo of my own which handles some customizations on top of your source.

So essentially my install looks a little something like this (on development):

```
FROM bitnami/wordpress
LABEL maintainer "Bitnami <containers@bitnami.com>"

# Install custom or additional server modules
RUN install_packages unzip nano vim

# Copy Over needed files
COPY scripts /

# Adjust script permissions
RUN chown root:root /customscript.sh
RUN chmod a+x /customscript.sh

# Expose Service Ports
EXPOSE 80 443

# Execute scripts
ENTRYPOINT [ "./customscript.sh" ]
CMD [ "nami", "start", "--foreground", "apache" ]
```

My customscript.sh just handles a few additional automated tasks, like installing a custom theme and plugins and activating them via wp-cli.

The issue I have as per my original post, we had a bug in a template that caused a php error.  So upon having this template installed with a php bug, any subsequent wp-cli command would output a php error (because of the template php bug) and that "error" response in the output would then cause the container to restart endlessly.

As soon as I manually fixed the template bug, the container wouldn't restart.  So that lead me to feel a bit uneasy about php bugs (during development) since they would show up on any wp-cli command during initialization which would then cause the container to restart endlessly.

That makes development more complex than it needs to IMO since I need to tamper with the initialization script and skip or fix on the fly these bugs before I could have a container up and running.  So they way I see this working more effectively is that PHP errors (during initialization) shouldn't prevent the container from staying up (that would make troubleshooting the actual php bug a lot easier and quicker).

Hopefully that makes sense.

Cheers

**Comment by sameersbn** [comment]
@momon Thanks for the information

> we had a bug in a template that caused a php error. So upon having this template installed with a php bug, any subsequent wp-cli command would output a php error

Now, things are making more sense. So basically the PHP script error renders the wp-cli command unstable which causes the `customscript.sh` to exit which makes k8s restart the pod.

> So they way I see this working more effectively is that PHP errors (during initialization) shouldn't prevent the container from staying up

Have you tried ignoring the errors in the customscript.sh? fe. the following will ignore any errors the wp-cli encounters. But this may not work as the `nami initialize` command may require that `wp-cli` to be working as expected.

```bash
wp-cli ... || true
```

> As soon as I manually fixed the template bug, the container wouldn't restart.

After fixing the PHP error you should:

1. rebuild the image with a new tag
2. update the image tag in the deployment 
3. `kubectl apply` the changes

The last step should remove the failed pod and replace it with a new one based on the new docker image. 

> That makes development more complex than it needs to IMO since I need to tamper with the initialization script and skip or fix on the fly these bugs before I could have a container up and running.

Honestly this image is not designed or intended for use in development scenarios as you can tell. 

You may want to take a look at https://www.telepresence.io/ for some ideas

**Comment by momon** [comment]
@sameersbn much appreciate the response and reply.  I have tried making wp-cli responses silent in numerous ways with built in params and via bash script tricks but they have no impact.

The process you have described is how I am working around the issue currently but that is somewhat poor solution to this kind of problem and I feel the actual problem is being dismissed without getting proper attention.  My case scenario above is just one of many and I described it as a "development issue" because I am under development myself, this however can be easily reproduced on any live environment which perhaps better showcases why this is an issue.

If you "accidentally" introduce any php error into wordpress, one common example is adding Google Adsense (or any other code) into functions.php file of any theme (which is a pretty common scenario), it will immediately break the website.  When the pod tries to rebuild, it will hit a PHP error during initialization and the pod will restart endlessly.

My point in case here and what I am trying to figure out (or suggest) is for a simpler way to IGNORE php errors as a condition to force a pod restart (at least on GKE which seems to ignore the pod restart policies in Kubernetes).

In simpler terms, an error in wordpress (PHP in this case) would need to be debugged while the pod is alive (at least under development).  I can understand why it would be undesirable to keep the pod off or restarting in production if the application is not working but even then it would be useful being able to keep the pod online despite PHP errors under some conditions.

Hopefully that makes sense.

I will give some things a try and if there are any other suggestions I'll be glad to look at them, I am sure others will come around similar issues sooner or later depending how they decide to use this image.

Thanks again!

**Comment by sameersbn** [comment]
@momon The thing I am trying to highlight is that the wordpress images is not designed for the development use case and whatever we come up with to make this work would be a workaround at best.  Like you said GKE seems to ignore the pod restart policies in Kubernetes, there is nothing we can do to address this.

**Comment by sameersbn** [comment]
@momon Like you said GKE seems to ignore the pod restart policies in Kubernetes, there is nothing we can do to address this.

The thing I am trying to highlight is that the wordpress images is not designed for the development use case and whatever we come up with to make this work would be a workaround at best. 

If you want the wordpress image to behave differently when PHP errors are encountered, you should 
 adapt the wordpress dockerfile for your requirements and rebuild the wordpress image.

If you just want the pod to be running, you can override the default entrypoint/command of the wordpress docker image in the wordpress chart to keep it running.

**Comment by momon** [comment]
@sameersbn awesome, that's in line with my intentions.  How would I invoke wordpress to start without nami then? So that the docker image is NOT depending on Wordpress / Apache / PHP to be working in order to remain online?

**Comment by sameersbn** [comment]
> How would I invoke wordpress to start without nami then?

**Step 1:**

Override the container launch command by patching the `deployment.yaml`

```diff
diff --git a/stable/wordpress/templates/deployment.yaml b/stable/wordpress/templates/deployment.yaml
index d24cfce8e..a534d71ec 100644
--- a/stable/wordpress/templates/deployment.yaml
+++ b/stable/wordpress/templates/deployment.yaml
@@ -44,6 +44,7 @@ spec:
       - name: wordpress
         image: {{ template "wordpress.image" . }}
         imagePullPolicy: {{ .Values.image.pullPolicy | quote }}
+        command: [ 'tail', '-f', '/dev/null']
         env:
         - name: ALLOW_EMPTY_PASSWORD
           value: {{ ternary "yes" "no" .Values.allowEmptyPassword | quote }}
```

**Step 2**:

Disable the readinessProbe while deploying the chart `--set readinessProbe.enabled=false`

At this point the wordpress chart is deployed but the application has not been launched yet!

**Step 3**:

Manually execute the wordpress containers start up sequence. If it fails due to PHP script errors, fix them and retry.

```bash
kubectl exec -it wp-wordpress-7fbf9469f6-4dx9s /app-entrypoint.sh -- nami start --foreground apache
```

I hope you find this information useful.

**Comment by momon** [comment]
@sameersbn while I much appreciate your efforts and guidance above, unfortunately it seems I have failed to understand exactly how to achieve your suggestions.

It may be due to the way I am deploying this as I have a single deployment file that uses this wordpress image and when I try adding the command you said in step 1 it seems to make no difference.  Reading through your comment it appears to me I should be modifying the image instead but I am unsure exactly how that is to be achieved exactly.

On step 2, based on a prior comment that suggested this, seems rather specific to helm, which I am NOT using for this type of setup.  Using regular deployments, as I provided an example above it was suggested that healthchecks are not being used:  https://github.com/bitnami/bitnami-docker-wordpress/issues/198#issuecomment-567911199

On step 3, I can understand the command but again, it assumes I am manually launching the pod, which is not the case.  I have a regular deployment which I am expecting to always be up and running regardless of PHP errors.

I do apologize if the above is perhaps dumb, I am relatively new to using docker but have been learning over the past few months.  Some concepts are not yet clear to me and I am trying to be resourceful.  This image I've been using heavily as it solves all my needs, I am just trying to figure out a way to prevent the image from restarting during php errors which is a critical part of what we are building.

**Comment by sameersbn** [comment]
@momon Oops, I forgot that you were not using Helm. 
But in any case you should add the line `command: [ 'tail', '-f', '/dev/null']` to your deployment yaml for the wordpress container. Step 2. is not required since probes are not used in your yaml. 

> On step 3, I can understand the command but again, it assumes I am manually launching the pod, which is not the case. I have a regular deployment which I am expecting to always be up and running regardless of PHP errors.

The pod should be running but it will not be running anything useful until you execute the specified command. If there are PHP error that command will fail, then you would have to fix the errors and execute the command again. Since you are using the container for development purposes, this is the only workaround I can offer you right now. I apologise if it does not meet your intended use.

**Comment by momon** [comment]
@sameersbn LEGEND! this is brilliant.  I tested it out and seems to work nearly as expected and makes perfect sense, so I certainly learned something here.

The one issue I still have, but I think I just might need to re-purpose the implementation, is keeping that command running.  When I executed via shell it obviously works but as soon as I exit the shell it kills the process.

So my understanding of what is happening is that the command in the deployment is meant at having a process that is "running" for the pod to remain active and we are sending that second command as an independent execution (which initializes wordpress).

That works for my intended use, what I now need to figure out is how to send that command over in a way that will keep the app running regardless of the shell.

**Comment by javsalgar** [comment]
My advice would be to mount a configmap with a custom script that covers your needs. In our chart we have the extraVolumes and extraVolumeMounts that could help you with this. Then you would change the command to use your script instead of the default command.

**Comment by momon** [comment]
Thanks again for the help @javsalgar, I did a few test of my own but it didn't quite work because as soon as I terminate the command, wordpress goes down obviously, but I think I can work around that by using a custom script like you said.  I will be running some tests and once I figured it out will be sure to update here in case someone else comes across a similar situation.

**Comment by sameersbn** [comment]
Happy to know that there was some progress here :+1: 
We would be happy to know the results of your investigation. Until then can we go ahead and close this issue?

**Comment by momon** [comment]
Thanks again for all the help.  I was able to do a lot of troubleshooting and while I do not yet have a solution, I have nailed this down to the actual culprit.  I am sharing this info here in case anyone has some ideas I could try as I haven't been able to fix it yet.

Upon debugging I was able to trace the constant restart to this particular portion of the script:

```
  nami_initialize apache php mysql-client wordpress
```

More specifically:

```
  nami_initialize wordpress
```

I purposely caused a PHP error to test by adding some lines into the functions.php file.  When nami_initialize wordpress runs after a pod restart, the pod reads this error from nami and exits:

```
nami    INFO  Initializing apache
nami    INFO  apache successfully initialized
nami    INFO  Initializing php
nami    INFO  php successfully initialized
nami    INFO  Initializing mysql-client
mysql-c INFO  Trying to connect to MySQL server
mysql-c INFO  Found MySQL server listening at mysql-sqlproxy-gcloud-sqlproxy.sqlproxy:3306
mysql-c INFO  MySQL server listening and working at mysql-sqlproxy-gcloud-sqlproxy.sqlproxy:3306
mysql-c INFO  ==> Found database dtb-fp-1-wp-4qsn9. Skipping database creation...
nami    INFO  mysql-client successfully initialized
nami    INFO  Initializing wordpress
wordpre INFO  ==> Preparing PHP environment...
wordpre INFO  WordPress has been already initialized, restoring...
mysql-c INFO  Trying to connect to MySQL server
mysql-c INFO  Found MySQL server listening at mysql-sqlproxy-gcloud-sqlproxy.sqlproxy:3306
mysql-c INFO  MySQL server listening and working at mysql-sqlproxy-gcloud-sqlproxy.sqlproxy:3306
wordpre INFO  Upgrading WordPress Database ...
Error executing 'postInstallation': PHP Notice:  Undefined index: HTTP_HOST in /bitnami/wordpress/wp-config.php on line 90
PHP Notice:  Undefined index: HTTP_HOST in /bitnami/wordpress/wp-config.php on line 91
PHP Parse error:  syntax error, unexpected 'if' (T_IF) in /bitnami/wordpress/wp-content/themes/twentynineteen/functions.php on line 16
```

After that syntax error, the pod restarts immediately and goes into a loop.

I was trying to look into nami_initialize but I don't seem to find any info on it.  I am trying to figure out how to:

1) Tell it to ignore these errors.
2) Manually implement in my custom image so I can control error handling

Some things I tried:

- Hiding all PHP errors from php.ini, it works as it doesn't show the PHP error, but it still restarts the pod.
- Tried adding a preStop hook to get the pod to stay alive for longer before restarting but it does not seem to have an impact in this scenario.

Now, I can make perfect sense that if nami can't get PHP working the app would "fail" but truth be told this is rather fragile.  It isn't far fetched that someone could accidentally update a plugin with an accidental syntax error.  I've seen it happen before at least twice.

There is also the change someone could be making live adjustments to the functions.php file which is rather common and accidentally insert bad syntax.

If this happens, then there is a bit of a grace period since the pod will not terminate, but based on scaling, as soon as the pod gets re-created, it will never start due to the php syntax error and go into an endless loop.

My ultimate goal is to have a longer window between restarts where I could execute some tasks that would allow me to recover the website.  That could be as simple as resetting the default theme or disabling all plugins so only the core files are running but with the pod constantly restarting, I don't have any sort of reasonable window to try and execute any operation to recover the pod.

Hopefully the above makes a bit more sense and someone is able to suggest something or point me in the right direction.

Cheers,

**Comment by sameersbn** [comment]
The `nami initialize` step uses the the command line interface for WordPress (`wp-cli`) to configure the wordpress installation. The PHP script errors are probably causing the `wp-cli` command to fail resulting in `nami initialize` failing. It's very critical for the `nami initialize` command to complete  successfully, else the wordpress app could be in a unusable state which is try it is treated as an error. 

Even if you later fix the error and launch the app again with the same colume, we cannot guarantee that the wordpress app  will be in a working state. 

Having said that if you still want to ignore the errors, then like I suggested before you need to modify the start up scripts to ignore the errors and build a custom image. 

For example  change https://github.com/bitnami/bitnami-docker-wordpress/blob/master/5/debian-10/rootfs/app-entrypoint.sh#L11 to `nami_initialize apache php mysql-client wordpress || true` and the `nami initialize` errors would be ignored.

I would reiterate that what you're trying to achieve is outside the scope of the design of this docker image.

**Comment by momon** [comment]
Thanks again @sameersbn and I completely understand.  I gave it a shot customizing the script as you suggested by using: ```nami_initialize apache php mysql-client wordpress || true``` but unfortunately that didn't work.  The script is still stuck at the PHP error and the pod restarts.

```
mysql-c INFO  Found MySQL server listening at mysql-sqlproxy-gcloud-sqlproxy.sqlproxy:3306
mysql-c INFO  MySQL server listening and working at mysql-sqlproxy-gcloud-sqlproxy.sqlproxy:3306
wordpre INFO  Upgrading WordPress Database ...
Error executing 'postInstallation': PHP Notice:  Undefined index: HTTP_HOST in /bitnami/wordpress/wp-config.php on line 90
PHP Notice:  Undefined index: HTTP_HOST in /bitnami/wordpress/wp-config.php on line 91
PHP Parse error:  syntax error, unexpected 'if' (T_IF) in /bitnami/wordpress/wp-content/themes/twentynineteen/functions.php on line 16
```

So I did a bit more research (https://make.wordpress.org/cli/handbook/config/) and it appears I can do --skip-plugins and/or --skip-themes on wp cli and that would effectively help me achieve my goals here, however, since these commands are within the actual script (nami_initialize wordpress) I don't have access to changing its behavior or adding the wp-cli parameters.

So while doing more testing I discovered the config file for wp-cli here:

```
/opt/bitnami/wp-cli/conf/wp-cli.yml
```

And using the following to append to the config file:

```
kubectl exec podxyz -- sed -i '$ a\skip-themes: true' /opt/bitnami/wp-cli/conf/wp-cli.yml
kubectl exec podxyz  -- sed -i '$ a\skip-plugins: true' /opt/bitnami/wp-cli/conf/wp-cli.yml
```

Allowed me to globally skip themes and plugins when needed which allows wp to initialize despite PHP errors.

So now I just need to wrap this into a custom function so I can use it selectively and enable a "debugging" mode on my pods.

Thanks a lot for your help and guidance, would not have been able to achieve this without your direction.

Much appreciate the support.

**Comment by sameersbn** [comment]
@momon Thanks really great. We will check internally if we can all the `--skip-plugins` and `--skip-themes` flags in our invocations of `wp-cli` tool. Thanks for your effort and patience.

**Comment by sameersbn** [comment]
putting this issued on-hold until we take a decision on whether we can add those args to our `wp-cli` invocations. will keep you posted. thanks again for the effort.

**Comment by momon** [comment]
@sameersbn that would actually be very beneficial for a lot more than just this purpose I am after.  I am in fact having trouble precisely there (despite being successfully in manual implementation as described above).

The main issue is that we have no control over the wp-cli configuration file while it isn't persisted and the folders where it would access it from based on the doc (so we can override the global config) are not persisted either.

Furthermore, while there is an env. variable that can be used to set the path of the config file that could be located on the persisted storage when using your image as a base (which is what I am using) I can't pass that environmental variable for the commands where I need it at since those are only loaded onto the final container (after your image has been processed).

So while I am still poking around for alternatives which I am sure I will sooner or later figure out, I think long term it might be very useful being able to setup wp-cli config options globally or persist a global wp-cli config file instead of using the default one that is provided so that we can have a bit more control over wp-cli regardless how the image is built.

What I am trying for the time being, is to use your image as is, setup my custom wp-cli config file on the first install with the settings I want and then have the build process use that config file instead of the global, however, I seem unable to get the base image to use my custom wp-cli config file because the paths where it would look for in priority are not persisted and I seem unable to pass the environmental variable to the FROM image instruction.

If I do figure out, I'll share my findings here, I just wanted to showcase why this would be relevant regardless my use case above, I think it would be very beneficial being able to control wp-cli config in a way that is persisted.

**Comment by momon** [comment]
Ok, I think I did now just found a bug as this doesn't appear to be intended behavior.

I was successful in doing both, setting up the custom config file for wp-cli on the root, confirmed to be working as per this output:

```
momon@DESKTOP-D8J28R5:~$ kubectl exec dpl-fp-1-wp-4qsn9-78b8b89b9-dhkrd -- wp --info --allow-root
OS:     Linux 4.19.76+ #1 SMP Tue Oct 8 23:17:06 PDT 2019 x86_64
Shell:
PHP binary:     /opt/bitnami/php/bin/php
PHP version:    7.3.14
php.ini used:   /opt/bitnami/php/lib/php.ini
WP-CLI root dir:        phar://wp-cli.phar/vendor/wp-cli/wp-cli
WP-CLI vendor dir:      phar://wp-cli.phar/vendor
WP_CLI phar path:       /
WP-CLI packages dir:
WP-CLI global config:   /opt/bitnami/wp-cli/conf/wp-cli.yml
WP-CLI project config:  //wp-cli.local.yml
WP-CLI version: 2.4.0
momon@DESKTOP-D8J28R5:~$ kubectl exec dpl-fp-1-wp-4qsn9-78b8b89b9-dhkrd -- cat //wp-cli.local.yml
# Global parameter defaults
path: "/opt/bitnami/wordpress/"
skip-themes: true
skip-plugins: true
```

Note that I was also able to edit the global config file before anything else starts in the script:

```
momon@DESKTOP-D8J28R5:~$ kubectl exec dpl-fp-1-wp-4qsn9-78b8b89b9-dhkrd -- cat /opt/bitnami/wp-cli/conf/wp-cli.yml
# Global parameter defaults
path: "/opt/bitnami/wordpress/"
skip-themes: true
skip-plugins: true
momon@DESKTOP-D8J28R5:~$
```

Both things take place at the very top of my custom init file as can be seen in the log output:

```
momon@DESKTOP-D8J28R5:~$ kubectl logs dpl-fp-1-wp-4qsn9-78b8b89b9-dhkrd

Welcome to the Bitnami wordpress container
Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-wordpress
Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-wordpress/issues
Send us your feedback at containers@bitnami.com

INFO  ==> Welcome to , we just need to setup some additional things...
INFO  ==> Setup wp-cli config file ...
INFO  ==> Setup wp-cli config completed
nami    INFO  Initializing apache
nami    INFO  apache successfully initialized
nami    INFO  Initializing php
nami    INFO  php successfully initialized
```

However, the wp commands within nami_initialize wordpress which are ran as part of the setup (according to earlier statements) seem to be ignoring both the project specific wp-cli config file as well as the global wp-cli config file based on this error:

```
momon@DESKTOP-D8J28R5:~$ kubectl logs dpl-fp-1-wp-4qsn9-78b8b89b9-dhkrd

Welcome to the Bitnami wordpress container
Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-wordpress
Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-wordpress/issues
Send us your feedback at containers@bitnami.com

INFO  ==> Welcome to , we just need to setup some additional things...
INFO  ==> Setup wp-cli config file ...
INFO  ==> Setup wp-cli config completed
nami    INFO  Initializing apache
nami    INFO  apache successfully initialized
nami    INFO  Initializing php
nami    INFO  php successfully initialized
nami    INFO  Initializing mysql-client
mysql-c INFO  Trying to connect to MySQL server
mysql-c INFO  Found MySQL server listening at mysql-sqlproxy-gcloud-sqlproxy.sqlproxy:3306
mysql-c INFO  MySQL server listening and working at mysql-sqlproxy-gcloud-sqlproxy.sqlproxy:3306
mysql-c INFO  ==> Found database dtb-fp-1-wp-4qsn9. Skipping database creation...
nami    INFO  mysql-client successfully initialized
nami    INFO  Initializing wordpress
wordpre INFO  ==> Preparing PHP environment...
wordpre INFO  WordPress has been already initialized, restoring...
mysql-c INFO  Trying to connect to MySQL server
mysql-c INFO  Found MySQL server listening at mysql-sqlproxy-gcloud-sqlproxy.sqlproxy:3306
mysql-c INFO  MySQL server listening and working at mysql-sqlproxy-gcloud-sqlproxy.sqlproxy:3306
wordpre INFO  Upgrading WordPress Database ...
Error executing 'postInstallation': PHP Notice:  Undefined index: HTTP_HOST in /bitnami/wordpress/wp-config.php on line 90
PHP Notice:  Undefined index: HTTP_HOST in /bitnami/wordpress/wp-config.php on line 91
PHP Parse error:  syntax error, unexpected 'if' (T_IF) in /bitnami/wordpress/wp-content/themes/twentynineteen/functions.php on line 16
```

To confirm, while I still have that theme broken, here is what running any other wp command on the container looks like:

```
momon@DESKTOP-D8J28R5:~$ kubectl exec dpl-fp-1-wp-4qsn9-78b8b89b9-dhkrd -- sudo -u daemon -- wp cron test
Warning: WP-Cron spawn succeeded but returned HTTP status code: 500 Internal Server Error
```

As you can see above, it does accept the settings from the config file and exits nicely (with just a warning) but the wp commands within the nami_initialize wordpress script to which I have no access, ignores the wp-cli settings.

Why would this be the case if not a bug? Unless the wp commands within that script have hardcoded values specific to that execution I don't see a reason they would ignore globally and app specific defined parameters.

Thanks!

**Comment by momon** [comment]
Here is a bit more info to further support why this issue is important.  I have Slider Revolution plugin installed (which comes bundled with the template I am using).  Slider Revolution is a widely known and commonly used Plugin.

It appears the version bundled with my theme (which is up to date) has some PHP warnings with this PHP version in the image and it shows warnings when executing any wp cli commands.  This is a widely used third party plugin I have no control over.

```
"PHP Warning: \"continue\" targeting switch is equivalent to \"break\". Did you mean to use \"continue 2\"? in /bitnami/wordpress/wp-content/plugins/slider-revolution/includes/operations.class.php on line 2858",

"PHP Warning: \"continue\" targeting switch is equivalent to \"break\". Did you mean to use \"continue 2\"? in /bitnami/wordpress/wp-content/plugins/slider-revolution/includes/operations.class.php on line 2858\nWarning: \"continue\" targeting switch is equivalent to \"break\". Did you mean to use \"continue 2\"? in /bitnami/wordpress/wp-content/plugins/slider-revolution/includes/operations.class.php on line 2858\nPHP Warning: \"continue\" targeting switch is equivalent to \"break\". Did you mean to use \"continue 2\"? in /bitnami/wordpress/wp-content/plugins/slider-revolution/includes/operations.class.php on line 2862\nWarning: \"continue\" targeting switch is equivalent to \"break\". Did you mean to use \"continue 2\"? in /bitnami/wordpress/wp-content/plugins/slider-revolution/includes/operations.class.php on line 2862"
```

While I was able to work around this by using --skip-themes and --skip-plugins on all our wp cli calls that clearly exposes why having a bit more control as suggested above is important.

**Comment by marcosbc** [comment]
Hi all, we have just released tag [`5.7.1-debian-10-r21`](https://github.com/bitnami/bitnami-docker-wordpress#571-debian-10-r21) of the container images which implements a lot of new features. In particular, it rewrites everything related to starting the Apache server with mod_php.

I'm going to close this issue as I haven't seen such error in new releases.

**Comment by marcosbc** [comment]
FYI `5.7.1-debian-10-r23` has also added support for specifying custom WP CLI flags (used during the initialization) via the `WORDPRESS_EXTRA_CLI_ARGS` environment variable, in which you can specify `--skip-plugins` and `--skip-themes`.
